pxr/usd/usdPhysics/rigidBodyAPI.cpp

```diff
@@ -250,3 +250,402 @@ PXR_NAMESPACE_CLOSE_SCOPE
+{
+    MassProperties massProps;
+
+    const UsdPrim usdPrim = GetPrim();
+    const UsdStagePtr stage = usdPrim.GetStage();
+
+    UsdGeomXformCache xfCache;
+
+    // Parse dynamic body mass data
+    const MassApiData rigidBodyMassInfo = ParseMassApi(usdPrim);
+
+    // If we dont have mass, we need to compute mass from collisions
+    const bool accumulateMass = rigidBodyMassInfo.mass <= 0.0f;
+
+    // Get initial data from parsed rigid body mass information
+    GfVec3f outCenterOfMass(0.0f);
+    const bool hasPa = rigidBodyMassInfo.hasPa;
+    GfQuatf outPrincipalAxes = rigidBodyMassInfo.principalAxes;
+    const bool hasCoM = GetCoM(usdPrim, outCenterOfMass, xfCache);    
+    float outMass = rigidBodyMassInfo.mass;
+    GfVec3f outDiagonalizedInertiaTensor = rigidBodyMassInfo.diagonalInertia;
+
+    // if we dont have enough mass information, we need to traverse collisions to gather final mass information
+    if (accumulateMass || !rigidBodyMassInfo.hasInertia || !hasCoM)
+    {
+        std::vector<MassProperties> massProps;
+        std::vector<GfMatrix4f> massTransf;
+        std::vector<UsdPrim> collisionPrims;        
+
+        // traverse all collisions below this body and get their collision information
+        // first gather all collisions
+        UsdPrimRange range(usdPrim);
+        for (auto it = range.begin(); it != range.end(); ++it)
+        {
+            const UsdPrim collisionPrim = *it;
+            if (collisionPrim && collisionPrim.HasAPI<UsdPhysicsCollisionAPI>())
+            {
+                collisionPrims.push_back(collisionPrim);
+            }
+        }
+
+        // get materials for all prims
+        std::vector<UsdShadeMaterial> physicsMaterials = UsdShadeMaterialBindingAPI::ComputeBoundMaterials(collisionPrims, _physicsPurposeTokens->materialPurposePhysics);
+        bool hasMaterials = true;
+        for (UsdShadeMaterial& material : physicsMaterials)
+        {
+            if (!material)
+            {
+                hasMaterials = false;                
+            }
+            else
+            {
+                if (!material.GetPrim().HasAPI<UsdPhysicsMaterialAPI>())
+                {
+                    hasMaterials = false;
+                    material = UsdShadeMaterial();
+                }
+            }
+        }
+
+        if (!hasMaterials)
+        {
+            std::vector<UsdShadeMaterial> generalMaterials = UsdShadeMaterialBindingAPI::ComputeBoundMaterials(collisionPrims);
+            for (size_t i = 0; i < physicsMaterials.size(); i++)
+            {
+                if (!physicsMaterials[i])
+                {
+                    const UsdShadeMaterial& generalMaterial = generalMaterials[i];
+                    if (generalMaterial && generalMaterial.GetPrim().HasAPI<UsdPhysicsMaterialAPI>())
+                    {
+                        physicsMaterials[i] = generalMaterial;
+                    }
+                }
+            }
+
+        }
```

Comment:
Ok, weird.  I added a comment in what must have been a forked branch version of this code that you don't need this second pass, here.